ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/Node/GSM.hs

```diff
@@ -0,0 +1,428 @@
+    -- file is absent.
+  }
+
+-----
+
+-- | Determine the initial 'LedgerStateJudgment'
+--
+-- Also initializes the persistent marker file.
+initializationLedgerJudgement ::
+     ( L.GetTip (L.LedgerState blk)
+     , Monad m
+     )
+  => m (L.LedgerState blk)
+  -> Maybe (WrapDurationUntilTooOld m blk)
+     -- ^ 'Nothing' if @blk@ has no age limit
+  -> MarkerFileView m
+  -> m LedgerStateJudgement
+initializationLedgerJudgement
+    getCurrentLedger
+    mbDurationUntilTooOld
+    markerFileView
+  = do
+    wasCaughtUp <- hasMarkerFile markerFileView
+    if not wasCaughtUp then pure TooOld else do
+        case mbDurationUntilTooOld of
+            Nothing -> return YoungEnough
+            Just wd -> do
+                sno <- L.getTipSlot <$> getCurrentLedger
+                getDurationUntilTooOld wd sno >>= \case
+                    After{}     -> return YoungEnough
+                    Already     -> do
+                        removeMarkerFile markerFileView
+                        return TooOld
+
+{-------------------------------------------------------------------------------
+  A real implementation
+-------------------------------------------------------------------------------}
+
+-- | The actual GSM logic for boot strap peers
+--
+-- See the @BootstrapPeersIER.md@ document for the specification of this logic.
+realGsmEntryPoints :: forall m blk upstreamPeer selection candidate.
+     ( MonadTimer m
+     , Eq upstreamPeer
+     )
+  => GsmView m blk upstreamPeer selection candidate
+  -> GsmEntryPoints m
+realGsmEntryPoints gsmView = GsmEntryPoints {
+    enterCaughtUp
+  ,
+    enterOnlyBootstrap
+  }
+  where
+    GsmView {
+        antiThunderingHerd
+      ,
+        candidateOverSelection
+      ,
+        durationUntilTooOld
+      ,
+        equivalent
+      ,
+        getChainSyncCandidates
+      ,
+        getChainSyncIdlers
+      ,
+        getCurrentSelection
+      ,
+        minCaughtUpDuration
+      ,
+        setCaughtUpPersistentMark
+      ,
+        varLedgerStateJudgement
+      } = gsmView
+
+    enterCaughtUp :: forall neverTerminates. m neverTerminates
+    enterCaughtUp  = enterCaughtUp' antiThunderingHerd
+
+    enterOnlyBootstrap :: forall neverTerminates. m neverTerminates
+    enterOnlyBootstrap  = enterOnlyBootstrap' antiThunderingHerd
+
+    enterCaughtUp' :: forall neverTerminates. StdGen -> m neverTerminates
+    enterCaughtUp' g = do
+        blockWhileCaughtUp
+
+        setCaughtUpPersistentMark False
+        enterOnlyBootstrap' g
+
+    enterOnlyBootstrap' :: StdGen -> forall neverTerminates. m neverTerminates
+    enterOnlyBootstrap' g = do
+        atomically $ StrictSTM.writeTVar varLedgerStateJudgement TooOld
+        blockUntilCaughtUp
+        atomically $ StrictSTM.writeTVar varLedgerStateJudgement YoungEnough
+
+        setCaughtUpPersistentMark True
+
+        -- When transitioning from OnlyBootstrap to CaughtUp, the node will
```

Comment:
I suddenly recalled discussing this logic with PNSol, but it wasn't in the IER.

What do you remember @amesgen? (I'll ask them either way.)

Reply:
I also remember that we talked about this in the context of avoiding "flickering" state transitions in case the tip of the honest chain is not very recent, which would induce undue load on the bootstrap peers.

We also explicitly listed the following advantage of the current transition rule for OnlyBootstrap->CaughtUp (see https://github.com/IntersectMBO/ouroboros-consensus/issues/101#issuecomment-1607767722):

> PRO: The trusted remote peer can also tell its downstream peer a "harmless lie" to make them conclude that they are caught up slightly prematurely. This might be especially useful during a disaster scenario (to relieve the trusted notes of load).

This of course only makes sense if the downstream peer actually stays in CaughtUp for at least some time and doesn't immediately switch back.

So I think it makes a lot of sense to have this "flickering-avoidance" mechanism.